### PR TITLE
Add --stats capabilities to display run statistics

### PR DESCRIPTION
**Pull request by egibs**
Closes: https://github.com/chainguard-dev/bincapz/issues/16

I wanted to try and address the goal(s) of the aforementioned issue.

This PR adds a new `--stats` flag which outputs the run's scan statistics at the end (outside of the active renderer).

I leveraged the recently-added `riskCounts` variable and added `packageRisks` as well. These are used to calculate the risk level percentages and risk namespace percentages, respectively.

I also made a few tweaks/improvements to `report.go` to clean up some of the logic and also exported the `ShortRisk` and `RiskLevels` variables to use them in `stats.go`.

Example(s):

File:
```
$ go run . --stats samples/macOS/2024.SpectralBlur.DPRK/SpectralBlur-macshare
Path: samples/macOS/2024.SpectralBlur.DPRK/SpectralBlur-macshare
Risk: 🔥 3/HIGH
--------------------------------------------------------------------------------------------------------------------------------------
meta    format                   macho
1/LOW   env/SHELL                users preferred SHELL path
1/LOW   exec/program/background  waits for a process to exit
1/LOW   fs/file/delete           deletes files
1/LOW   fs/symlink/resolve       resolves symbolic links
1/LOW   net/hostname/resolve     resolves network hosts via name
1/LOW   net/socket/receive       receive a message from a socket
1/LOW   net/socket/send          send a message to a socket
1/LOW   process/create           create a new child process using fork
1/LOW   process/multithreaded    uses pthreads
1/LOW   process/username/get     get login name
1/LOW   random/insecure          generate random numbers insecurely
2/MED   device/pseudo_terminal   pseudo-terminal access functions
2/MED   exec/program             executes another program
2/MED   kernel/uname/get         get system identification (uname)
2/MED   net/download             downloads files
2/MED   net/ip/parse             parses IP address
2/MED   net/ip/string            converts IP address from byte to string
2/MED   net/socket/connect       initiate a connection on a socket
2/MED   net/upload               uploads files
2/MED   shell/exec               executes a shell
3/HIGH  combo/backdoor/net_term  uploads, provides a terminal, runs program: "_uname _unlink _waitpid execve shell tcsetattr upload"

📊 Statistics
---
⚠️  Risk Level Percentage
---
Risk Level   Percentage
3/HIGH           100.00%
---
📦 Package Risk Percentage
---
Package                 Percentage
net/socket/receive            8.33%
net/socket/send               8.33%
random/insecure               4.17%
exec/program/background       4.17%
shell/exec                    4.17%
process/username/get          4.17%
device/pseudo_terminal        4.17%
net/ip/string                 4.17%
fs/file/delete                4.17%
kernel/uname/get              4.17%
process/multithreaded         4.17%
fs/symlink/resolve            4.17%
net/upload                    4.17%
process/create                4.17%
net/socket/connect            4.17%
combo/backdoor/net_term       4.17%
exec/program                  4.17%
net/hostname/resolve          4.17%
net/download                  4.17%
env/SHELL                     4.17%
net/ip/parse                  4.17%
meta/format/macho             4.17%
```

Full (only showing statistics for brevity):
```
📊 Statistics
---
⚠️  Risk Level Percentage
---
Risk Level   Percentage
0/NONE            93.79%
2/MED              3.10%
3/HIGH             1.43%
4/CRIT             0.95%
1/LOW              0.72%
---
📦 Package Risk Percentage
---
Package                                  Percentage
ref/site/url                                   3.76%
shell/exec                                     3.27%
exec/program                                   2.29%
net/socket/send                                2.13%
net/socket/receive                             1.96%
fs/permission/modify                           1.80%
encoding/base64                                1.80%
ref/path/usr/bin                               1.47%
ref/path/tmp                                   1.47%
ref/path/etc                                   1.47%
net/http/post                                  1.47%
kernel/uname/get                               1.31%
process/multithreaded                          1.15%
ref/path/hidden                                1.15%
meta/lang/c++                                  1.15%
net/fetch                                      1.15%
fs/directory/create                            0.98%
kernel/hostname/get                            0.98%
fs/file/delete                                 0.98%
fs/file/times/set                              0.98%
net/socket/listen                              0.98%
ref/path/var                                   0.82%
net/socket/local/address                       0.82%
crypto/aes                                     0.82%
net/url/request                                0.82%
meta/format/macho                              0.82%
net/http_proxy                                 0.82%
net/url                                        0.82%
net/ip/parse                                   0.82%
combo/recon/system_network                     0.82%
encoding/json                                  0.82%
process/username/get                           0.82%
exec/program/background                        0.65%
shell/ignore_output                            0.65%
fs/file/read                                   0.65%
kernel/cpu/info                                0.65%
net/http/request                               0.65%
systemd/out_of_dependency_tree                 0.65%
ref/path/relative                              0.65%
net/socket/connect                             0.65%
fs/link/read                                   0.65%
fs/directory/remove                            0.65%
env/TERM                                       0.65%
ref/words/password                             0.65%
archives/zip                                   0.65%
fs/tempfile/create                             0.65%
fs/permission/chown                            0.65%
combo/dropper/bash                             0.65%
net/hostport/parse                             0.65%
random/insecure                                0.65%
net/ip/string                                  0.65%
fs/file/make_executable                        0.65%
process/chdir                                  0.65%
process/create                                 0.65%
ref/path/var/tmp                               0.65%
net/socket/peer/address                        0.65%
systemd/no_blank_lines                         0.65%
net/hostname/resolve                           0.65%
systemd/execstart/elsewhere                    0.49%
net/http/cookies                               0.49%
compression/gzip                               0.49%
env/USER                                       0.49%
net/interface/list                             0.49%
ref/path/etc/resolv.conf                       0.49%
techniques/code_eval                           0.49%
persist/crontab                                0.49%
ref/ip/dns_resolver                            0.49%
net/http/auth                                  0.49%
data/embedded/base64/url                       0.49%
3P/signature_base/redmenshen/bpfdoor           0.49%
fs/directory/list                              0.49%
process/userid/set                             0.49%
ref/site/http/ip                               0.49%
evasion/base64/external/decoder                0.49%
systemd/no_docs_or_comments                    0.49%
net/http/accept/encoding                       0.49%
net/dns/txt                                    0.49%
evasion/base64/eval                            0.49%
fs/lock/update                                 0.49%
net/download                                   0.49%
encoding/json/decode                           0.49%
encoding/json/encode                           0.49%
ref/words/trojan                               0.33%
net/icmp                                       0.33%
sync/semaphore/user                            0.33%
evasion/rename_system_binary                   0.33%
evasion/base64/http                            0.33%
process/groups/set                             0.33%
net/udp/receive                                0.33%
net/dns                                        0.33%
fs/file/delete/forcibly                        0.33%
process/groupid/set                            0.33%
env/HOME                                       0.33%
net/url/encode                                 0.33%
crypto/tls                                     0.33%
dylib/symbol/address                           0.33%
fs/file/truncate                               0.33%
ref/words/intercept                            0.33%
exec/shell_command                             0.33%
ref/words/agent                                0.33%
compression/zstd                               0.33%
net/ip/multicast/send                          0.33%
data/embedded/pgp/key                          0.33%
device/pseudo_terminal                         0.33%
systemd/restart/always                         0.33%
net/upload                                     0.33%
kernel/dispatch/semaphore                      0.33%
ref/path/etc/hosts                             0.33%
security_controls/linux/iptables               0.33%
process/chroot                                 0.33%
evasion/base64/shell/commands                  0.33%
net/sendfile                                   0.33%
net/udp/send                                   0.33%
persist/launch/agent                           0.33%
ref/path/home                                  0.33%
crypto/ecdsa                                   0.33%
data/embedded/pem/private_key                  0.33%
admin/add_apt_key                              0.33%
exec/pipe                                      0.33%
admin/system_directories                       0.33%
secrets/keychain                               0.33%
combo/backdoor/net_term                        0.33%
secrets/private_key                            0.33%
kernel/sysctl/nmi_watchdog                     0.33%
process/name/set                               0.33%
net/interface/get                              0.33%
meta/compiler/go                               0.33%
ref/ip                                         0.33%
meta/entitlements                              0.33%
evasion/content/length/0                       0.33%
net/http2                                      0.33%
ref/words/infected                             0.33%
net/reuseport                                  0.33%
ref/program/sshd                               0.33%
fd/multiplex                                   0.33%
crypto/ed25519                                 0.33%
fd/epoll                                       0.16%
security_controls/linux/selinux_disable        0.16%
env/SHELL                                      0.16%
ref/program/masscan                            0.16%
net/fetch/suspicious                           0.16%
ref/path/usr/sbin                              0.16%
shell/bash_dev_tcp                             0.16%
combo/backdoor/sys_cmd                         0.16%
combo/net/tunnel_proxy                         0.16%
combo/degrader/selinux_firewall                0.16%
service/systemd                                0.16%
3P/elastic/bpfdoor                             0.16%
exec/shell_echo                                0.16%
kernel/netlink                                 0.16%
crypto/mining/generic                          0.16%
ref/program/sudo                               0.16%
process/find                                   0.16%
fs/symlink/resolve                             0.16%
3P/signature_base/nk/3cx                       0.16%
dylib/address/check                            0.16%
evasion/hide_shell_history                     0.16%
ref/path/usr/local                             0.16%
tty/getpass                                    0.16%
net/fetch/insecure                             0.16%
kernel/sysinfo                                 0.16%
net/geoip                                      0.16%
data/embedded/base64/terms                     0.16%
security_controls/linux/ufw                    0.16%
device/disk/info                               0.16%
ref/words/dropper                              0.16%
3P/signature_base/susp/xored                   0.16%
kernel/hardware/info                           0.16%
evasion/base64/decode                          0.16%
evasion/bash_tcp                               0.16%
net/raw_sockets                                0.16%
combo/net/raw_flooder                          0.16%
ref/program/ancient_gcc                        0.16%
3P/signature_base/3cxdesktopapp/backdoor       0.16%
combo/backdoor/py_setuptools                   0.16%
net/ip/send/unicast                            0.16%
process/list                                   0.16%
data/embedded/html                             0.16%
combo/stealer/upload/keychain/zip              0.16%
net/dns/over/https                             0.16%
security_controls/linux/selinux                0.16%
fs/watch                                       0.16%
tty/vhangup                                    0.16%
ref/program/nmap                               0.16%
procfs/arbitrary/pid                           0.16%
compression/bzip2                              0.16%
shell/arbitrary_command/dev_null               0.16%
meta/lang/objc                                 0.16%
fs/directory/traverse                          0.16%
net/ip/byte/order                              0.16%
systemd/short/description                      0.16%
evasion/codecs_decode                          0.16%
hash/sha1                                      0.16%
net/public_ip                                  0.16%
3P/signature_base/pua/crypto                   0.16%
3P/godmoderules/iddqd/god/mode                 0.16%
meta/program_name                              0.16%
service/stop                                   0.16%
data/embedded/pem/test_key                     0.16%
evasion/base64/python                          0.16%
3P/volexity/iconic                             0.16%
kernel/ptrace                                  0.16%
```

Without `--stats` to show that the flag works:
```
❯ go run . samples/macOS/2024.SpectralBlur.DPRK/SpectralBlur-macshare
Path: samples/macOS/2024.SpectralBlur.DPRK/SpectralBlur-macshare
Risk: 🔥 3/HIGH
--------------------------------------------------------------------------------------------------------------------------------------
meta    format                   macho
1/LOW   env/SHELL                users preferred SHELL path
1/LOW   exec/program/background  waits for a process to exit
1/LOW   fs/file/delete           deletes files
1/LOW   fs/symlink/resolve       resolves symbolic links
1/LOW   net/hostname/resolve     resolves network hosts via name
1/LOW   net/socket/receive       receive a message from a socket
1/LOW   net/socket/send          send a message to a socket
1/LOW   process/create           create a new child process using fork
1/LOW   process/multithreaded    uses pthreads
1/LOW   process/username/get     get login name
1/LOW   random/insecure          generate random numbers insecurely
2/MED   device/pseudo_terminal   pseudo-terminal access functions
2/MED   exec/program             executes another program
2/MED   kernel/uname/get         get system identification (uname)
2/MED   net/download             downloads files
2/MED   net/ip/parse             parses IP address
2/MED   net/ip/string            converts IP address from byte to string
2/MED   net/socket/connect       initiate a connection on a socket
2/MED   net/upload               uploads files
2/MED   shell/exec               executes a shell
3/HIGH  combo/backdoor/net_term  uploads, provides a terminal, runs program: "_uname _unlink _waitpid execve shell tcsetattr upload"


```

If I missed anything or if there's any feedback I'd be more than happy to address it!

`make lint`, `make fix`, and `make test` are all good as well.

### CODE REVIEW

**Comment by tstromberg** [comment]
At first glance, this looks super fantastic! Thank you for contributing this improvement. I'll give it a more thorough review soon.

File: pkg/render/stats.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+func riskLevelStatistics(files map[string]bincapz.FileReport, riskMap map[int][]string, riskPercentages map[int]float64) []bincapz.Kv {
+	for path, rf := range files {
+		riskMap[rf.RiskScore] = append(riskMap[rf.RiskScore], path)
+		for riskLevel := range riskMap {
+			riskPercentages[riskLevel] = (float64(len(riskMap[riskLevel])) / float64(len(files))) * 100
+		}
+	}
+
+	var riskLevelStatistics []bincapz.Kv
```
Comment (tstromberg) [inline pkg/render/stats.go:19]: Can you use a shorter name for this? Maybe `rls` or `stats`? 

In Go, the length of a variable name indicates the size of its scope, and this variable is only used for 8 lines. The name is also the same as the function it resides in, which is unusual and causes a stutter pattern: https://go.dev/wiki/CodeReviewComments#variable-names

File: pkg/render/stats.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	sort.Slice(riskLevelStatistics, func(i, j int) bool {
+		return riskLevelStatistics[i].Value > riskLevelStatistics[j].Value
+	})
+
+	return riskLevelStatistics
+}
+
+func packageRiskStatistics(files map[string]bincapz.FileReport, packageRiskMap map[string]int) ([]bincapz.KvStr, int) {
+	var numNamespaces int
+	packageRiskPercentages := make(map[string]float64)
```
Comment (tstromberg) [inline pkg/render/stats.go:32]: Same here, this could easily just be "percentages" or "stats"

File: pkg/bincapz/bincapz.go [hunk 3]
```diff
@@ -53,3 +54,13 @@ type Report struct {
 	Diff   DiffReport            `json:",omitempty" yaml:",omitempty"`
 	Filter string                `json:",omitempty" yaml:",omitempty"`
 }
+
+type Kv struct {
```
Comment (tstromberg) [inline pkg/bincapz/bincapz.go:58]: Can you pick a name that makes its use more obvious?

I was thinking something like "MetricInt" or "StatInt" but I'm sure there are better ideas here.

**Review by tstromberg — CHANGES_REQUESTED** [review]

**Comment by tstromberg** [comment]
Looks great - no problems with the code, just have some naming suggestions.

File: pkg/render/stats.go (hunk 1 above)
Comment (egibs) [reply pkg/render/stats.go]: Appreciate the context!

Addressed in [`5ad17ab` (#121)](https://github.com/chainguard-dev/bincapz/pull/121/commits/5ad17abd03cd920455ed146562cb0bf3c9ac992f).

File: pkg/render/stats.go (hunk 2 above)
Comment (egibs) [reply pkg/render/stats.go]: Addressed in [`5ad17ab` (#121)](https://github.com/chainguard-dev/bincapz/pull/121/commits/5ad17abd03cd920455ed146562cb0bf3c9ac992f).

File: pkg/bincapz/bincapz.go (hunk 3 above)
Comment (egibs) [reply pkg/bincapz/bincapz.go]: Addressed in [`5ad17ab` (#121)](https://github.com/chainguard-dev/bincapz/pull/121/commits/5ad17abd03cd920455ed146562cb0bf3c9ac992f).

**Review by tstromberg — APPROVED** [review]
Looks amazing - I can't wait to use this for gathering false positive data.

Thank you for contributing to bincapz!

**Pull request merged**